src/qml_benchmarks/models/base.py

```diff
@@ -0,0 +1,299 @@
+        # Therefore we add it here for clarity.
+        self.history_: list[float] = None
+        self.training_time_: float = None
+
+        self.mcmc_step = jax.jit(self.mcmc_step)
+        self.batched_mcmc_sample = jax.vmap(
+            self.mcmc_sample, in_axes=(None, 0, None, 0)
+        )
+
+    def generate_key(self):
+        return jax.random.PRNGKey(self.rng.integers(1000000))
+
+    @abstractmethod
+    def energy(self, params: dict, x: any) -> float:
+        """
+        The energy function for the model for a given configuration x.
+
+        This function should be implemented by the subclass and is also
+        responsible for initializing the parameters of the model, if necessary.
+
+        Args:
+            x: The configuration to calculate the energy for.
+        Returns:
+            energy (float): The energy.
+        """
+        pass
+
+    # TODO: this can be made even more efficient with Numpyro MCMC
+    # see qgml.mcmc for a simple example
+    def mcmc_step(self, args, i):
+        """
+        Perform one metropolis hastings steps.
+        The format is such that it can be used with jax.lax.scan for fast compilation.
+        """
+        params, key, x = args
+        key1, key2 = jax.random.split(key, 2)
+        flip_idx = jax.random.choice(key1, jnp.arange(self.dim))
+        flip_config = jnp.zeros(self.dim, dtype=int)
+        flip_config = flip_config.at[flip_idx].set(1)
+        x_flip = jnp.array((x + flip_config) % 2)
+
+        en = self.energy(params, jnp.expand_dims(x, 0))[0]
+
+        en_flip = self.energy(params, jnp.expand_dims(x_flip, 0))[0]
+        accept_ratio = jnp.exp(-en_flip) / jnp.exp(-en)
+        accept = jnp.array(jax.random.bernoulli(key2, accept_ratio), dtype=int)[0]
+        x_new = accept * x_flip + (1 - accept) * x
+        return [params, key2, x_new], x
+
+    def mcmc_sample(self, params, x_init, num_mcmc_steps, key):
+        """
+        Sample a chain of configurations from a starting configuration x_init
+        """
+        carry = [params, key, x_init]
+        carry, configs = jax.lax.scan(self.mcmc_step, carry, jnp.arange(num_mcmc_steps))
+        return configs
+
+    def langevin_sample(self, params, x_init, n_samples, key):
+        pass
+
+    def sample(self, num_samples, num_mcmc_steps=1000):
+        """
+        sample configurations starting from a random configuration.
+        """
+        if self.params_ is None:
+            raise ValueError(
+                "Model not initialized. Call model.initialize first with"
+                "example data sample."
+            )
+        keys = jax.random.split(self.generate_key(), num_samples)
+
+        # we do not take the gradient wrt the sampling, so decouple the param dict here
+        params_copy = copy.deepcopy(self.params_)
+
+        for key in params_copy.keys():
+            params_copy[key] = jax.lax.stop_gradient(params_copy[key])
```

Comment:
I don't think we need to copy the params here because we aren't taking the gradient. So you can just use self.params_